### ISSUE TITLE
Very Strange Triangulation of Segmentation, Cannot Globally Update Marker Size or Face Color

### ISSUE THREAD

**Issue by jeffquinn-msk** [scverse/napari-spatialdata#195]
I tried out napari-spatialdata on a Xenium project and I noticed a few usability issues. I wonder if these are specific to me (maybe my workstation doesn't have sufficient resources? I have an M1 macbook with 16GB memory). Overall the program was very laggy/non-responsive so that makes me suspect a resource limitation.

The first thing is that the triangulation of the segmentation shapes seems kind of wonky. This is not at all what the nuclear segments look like in the Xenium browser. Is there some configuration parameter I can pass to get a higher poly rendering of the segments?

After that there were a few usability issues, like not being able to change the size of the transcript dots globally, however adjusting the opacity globally seemed to work fine.

https://github.com/scverse/napari-spatialdata/assets/105746884/3036634e-91b9-422c-9d51-9821bdcfad2d

It seems I need to actually select the dots to change their size, but when I try and use cmd-A to select them all the software crashed (hung indefinitely). 

https://github.com/scverse/napari-spatialdata/assets/105746884/3919c69d-4a6f-47e1-b0de-19d8efef69c0

Maybe these issues are not relevant to this software project directly but I'm just curious if something is horribly wrong with my installation or if this is expected. If I'm viewing a full Xenium slide should I maybe downsample to just one quadrant of the image maybe?

**Comment by LucaMarconato** [comment]
Hi @jeffquinn-msk thanks for reporting this. I'll answer point-by-point.

> The first thing is that the triangulation of the segmentation shapes seems kind of wonky. This is not at all what the nuclear segments look like in the Xenium browser. Is there some configuration parameter I can pass to get a higher poly rendering of the segments?

@jeffquinn-msk to double check if this is a problem of napari and not of the Xenium dataset you are using, please plot single polygons using spatialdata-plot or the shapely plotting APIs. Xenium Explorer reads the polygons from the `cells.zarr.zip` file not from the `.parquet` files, so it could be that they are different.

If the above is not the case, @melonora do you know why the downsampling happens? I have never noticed this. We discussed adding this option in `napari-spatialdata` to mitigate the poor performance with shapes but I don't recall someone adding it in the end, so maybe this is happening at the napari side?

> After that there were a few usability issues, like not being able to change the size of the transcript dots globally, however adjusting the opacity globally seemed to work fine.

This is the napari design, I also found it a bit confusing in the beginning but then I got used. If you could open a issue on napari on this I will also jump in the conversation please.

> It seems I need to actually select the dots to change their size, but when I try and use cmd-A to select them all the software crashed (hung indefinitely).

For me this works, but selecting lot of points takes some time (napari limitation), I have a machine with higher specs so in my case it's still acceptable. I would consider reporting also this to napari please.

> Maybe these issues are not relevant to this software project directly but I'm just curious if something is horribly wrong with my installation or if this is expected. If I'm viewing a full Xenium slide should I maybe downsample to just one quadrant of the image maybe?

For me even if points are not a problem, I can't use napari for rendering the full set of polygons if I first don't crop the data or remove vertices. This has been reported to napari before but unfortunately they currently don't have the bandwidth to work on this. We are exploring internally some ideas to mitigate this as part of the `napari-spatialdata` plugin, such as using multiscale vector representations or rasterizing the vector data.

We have now an updated version of the `xenium()` reader that reads the multiscale raster masks for the cells and nuclei. Maybe this would be useful for your use case (the table needs to be mapped to the labels before being able to plot the annotations). https://github.com/scverse/spatialdata-io/pull/112

**Comment by melonora** [comment]
Hi @jeffquinn-msk,

A few of the issues you are having are seemingly related to parts of napari-core. We have a class called `EventedSet` that could be slowing down things when doing very large scale manipulations. Currentlly work is in progress to fix that.

Regarding the wonky segmentations. I wonder if this is due to the `epsilon` parameter of the rdp algorithm. In the `napari` menubar could you please go to `File` -> `preferences` and then the `Experimental` tab. Here you see the `Shapes polygon lasso RDP epsilon` parameter. Set this to 0 and please check if it resolves the issue. If not, I would have happy to jump in a call to try get a better understanding of your issue so we can resolve it.

**Comment by melonora** [comment]
Yes this was indeed related to 'EventedSet' which uses that library
